File: python-algorithm/leetcode/problem_494.py
"""494. Target Sum
https://leetcode.com/problems/target-sum/

You are given a list of non-negative integers, a1, a2, ..., an, and a target,
S. Now you have 2 symbols + and -. For each integer, you should choose one
from + and - as its new symbol.

Find out how many ways to assign symbols to make sum of integers equal to
target S.

Example 1:

Input: nums is [1, 1, 1, 1, 1], S is 3.
Output: 5
Explanation:

-1+1+1+1+1 = 3
+1-1+1+1+1 = 3
+1+1-1+1+1 = 3
+1+1+1-1+1 = 3
+1+1+1+1-1 = 3

There are 5 ways to assign symbols to make the sum of nums be target 3.

Note:

The length of the given array is positive and will not exceed 20.
The sum of elements in the given array will not exceed 1000.
Your output answer is guaranteed to be fitted in a 32-bit integer.
"""
from functools import lru_cache
from typing import List


class Solution:
    def brute_force(self, nums: List[int], s: int) -> int:
        def dfs(idx: int, target: int):
            max_sum = sum(nums[idx:])
            if target > max_sum or target < -max_sum:
                return
            if idx == len(nums) - 1:
                nonlocal ways
                if nums[idx] == target:
                    ways += 1
                if nums[idx] == -target:
                    ways += 1
                return
            dfs(idx + 1, target - nums[idx])
            dfs(idx + 1, target + nums[idx])

        ways = 0
        dfs(0, s)
        return ways

    def dfs(self, nums: List[int], s: int) -> int:
        if not nums:
            return 1 if s == 0 else 0
        ways = 0
        ways += self.dfs(nums[1:], s - nums[0])
        ways += self.dfs(nums[1:], s + nums[0])
        return ways

    def find_target_sum_ways(self, nums: List[int], s: int) -> int:
        # assume dp[i][j] as the total ways of nums[0:i] to sum j.
        # dp[i][j] = dp[i - 1][j - nums[i]] + dp[i - 1][j + nums[i]]
        length = len(nums)
        dp = [[0] * 2001 for _ in range(length)]
        dp[0][1000 + nums[0]] += 1
        dp[0][1000 - nums[0]] += 1
        for i in range(1, length):
            for j in range(-1000, 1001):
                if dp[i - 1][1000 + j] > 0:
                    dp[i][1000 + j + nums[i]] += dp[i - 1][1000 + j]
                    dp[i][1000 + j - nums[i]] += dp[i - 1][1000 + j]
        return dp[length - 1][1000 + s] if abs(s) <= 1000 else 0

    def find_target_sum_ways2(self, nums: List[int], target: int) -> int:
        @lru_cache(None)
        def dfs(i: int, t: int) -> int:
            if i == len(nums):
                return 1 if t == 0 else 0
            return dfs(i + 1, t - nums[i]) + dfs(i + 1, t + nums[i])

        return dfs(0, target)

    def find_target_sum_ways3(self, nums: List[int], target: int) -> int:
        offset = sum(nums)
        if target > offset:
            return 0
        limit = 2 * offset + 1
        dp = [[0] * limit for _ in range(len(nums))]
        if nums[-1] == 0:
            dp[-1][offset - nums[-1]] = 2
        else:
            dp[-1][offset - nums[-1]] = 1
            dp[-1][offset + nums[-1]] = 1
        for i in range(len(nums) - 2, -1, -1):
            for j in range(-offset, offset + 1):
                if offset + j >= 0:
                    dp[i][offset + j] += dp[i + 1][offset + j - nums[i]]
                if offset + j + nums[i] < limit:
                    dp[i][offset + j] += dp[i + 1][offset + j + nums[i]]
        return dp[0][offset + target]

File: java-algo/src/main/java/com/leetcode/Problem494.java
package com.leetcode;

import java.util.Arrays;

/**
 * 494. Target Sum
 * https://leetcode.com/problems/target-sum/
 *
 * You are given a list of non-negative integers, a1, a2, ..., an, and a target,
 * S. Now you have 2 symbols + and -. For each integer, you should choose one
 * from + and - as its new symbol.
 * ⁠
 *
 * Find out how many ways to assign symbols to make sum of integers equal to
 * target S.
 *
 *
 * Example 1:
 *
 * Input: nums is [1, 1, 1, 1, 1], S is 3.
 * Output: 5
 * Explanation:
 *
 * -1+1+1+1+1 = 3
 * +1-1+1+1+1 = 3
 * +1+1-1+1+1 = 3
 * +1+1+1-1+1 = 3
 * +1+1+1+1-1 = 3
 *
 * There are 5 ways to assign symbols to make the sum of nums be target 3.
 *
 *
 * Note:
 *
 * The length of the given array is positive and will not exceed 20.
 * The sum of elements in the given array will not exceed 1000.
 * Your output answer is guaranteed to be fitted in a 32-bit integer.
 */
public class Problem494 {

    /**
     * Time Complexity: O(2^n)
     * Space Complexity: O(n)
     */
    public int findTargetSumWays1(int[] nums, int s) {
        int length = nums.length;
        if (length == 0) return s == 0 ? 1 : 0;
        int[] nextArr = Arrays.copyOfRange(nums, 1, length);
        return findTargetSumWays1(nextArr, s - nums[0]) + findTargetSumWays1(nextArr, s + nums[0]);
    }

    private int ways = 0;
    public int findTargetSumWays2(int[] nums, int s) {
        recurse(nums, 0, 0, s);
        return ways;
    }

    private void recurse(int[] nums, int idx, int sum, int s) {
        if (idx == nums.length) {
            if (sum == s) ways++;
            return;
        }
        recurse(nums, idx + 1, sum + nums[idx], s);
        recurse(nums, idx + 1, sum - nums[idx], s);
    }

    public int findTargetSumWays3(int[] nums, int s) {
        // there're 2001 numbers between [-1000, 1000]
        int[][] store = new int[nums.length][2001];
        for (int[] row : store) {
            Arrays.fill(row, Integer.MIN_VALUE);
        }
        return recurse2(nums, 0, 0, s, store);
    }

    private int recurse2(int[] nums, int idx, int sum, int s, int[][] store) {
        if (idx == nums.length) {
            return sum == s ? 1 : 0;
        }
        if (store[idx][sum + 1000] != Integer.MIN_VALUE) {
            return store[idx][sum + 1000];
        }
        int ways = recurse2(nums, idx + 1, sum + nums[idx], s, store) +
                recurse2(nums, idx + 1, sum - nums[idx], s, store);
        store[idx][sum + 1000] = ways;
        return ways;
    }

    public int findTargetSumWays4(int[] nums, int s) {
        int length = nums.length;
        // dp[i][j] represents the total ways to sum j by nums[0:i].
        int[][] dp = new int[length][2001];
        return dp[length - 1][s + 1000];
    }

}
